### PR TITLE
Fix: Guard against missing names in AST

### PR DESCRIPTION
**Pull request by ekosz**
So another weird problem I've encountered with a hacky fix. This is the error I'm getting:

```
Error: Cannot read property 'value' of undefined
```

This is what my query looks like to cause this:

``` graphql
{
  checkin(id: 9) {
    entityId
    createdAt
    relationship {
      manager {
        ...UserDetails
      }
      report {
        ...UserDetails
        email
        company {
          slug
        }
      }
    }
    answers {
      edges {
        node {
          body(format: HTML)
          comments {
            edges {
              node {
                ... on CheckinAnswerComment {
                  body(format: HTML)
                  commenter: user {
                    name
                    avatarUrl
                  }
                }
              }
            }
          }
          question {
            body
          }
        }
      }
    }
  }
}

fragment UserDetails on User {
  name
  avatarUrl
}
```

And the ast object looks like:

``` json
{
  "kind": "InlineFragment",
  "typeCondition": {
    "kind": "NamedType",
    "name": {
      "kind": "Name",
      "value": "CheckinAnswerComment",
      "loc": {
        "start": 380,
        "end": 400,
        "source": {
          "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
          "name": "GraphQL request"
        }
      }
    },
    "loc": {
      "start": 380,
      "end": 400,
      "source": {
        "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
        "name": "GraphQL request"
      }
    }
  },
  "directives": [

  ],
  "selectionSet": {
    "kind": "SelectionSet",
    "selections": [
      {
        "kind": "Field",
        "alias": null,
        "name": {
          "kind": "Name",
          "value": "body",
          "loc": {
            "start": 421,
            "end": 425,
            "source": {
              "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
              "name": "GraphQL request"
            }
          }
        },
        "arguments": [
          {
            "kind": "Argument",
            "name": {
              "kind": "Name",
              "value": "format",
              "loc": {
                "start": 426,
                "end": 432,
                "source": {
                  "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                  "name": "GraphQL request"
                }
              }
            },
            "value": {
              "kind": "EnumValue",
              "value": "HTML",
              "loc": {
                "start": 434,
                "end": 438,
                "source": {
                  "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                  "name": "GraphQL request"
                }
              }
            },
            "loc": {
              "start": 426,
              "end": 438,
              "source": {
                "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                "name": "GraphQL request"
              }
            }
          }
        ],
        "directives": [

        ],
        "selectionSet": null,
        "loc": {
          "start": 421,
          "end": 439,
          "source": {
            "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
            "name": "GraphQL request"
          }
        }
      },
      {
        "kind": "Field",
        "alias": {
          "kind": "Name",
          "value": "commenter",
          "loc": {
            "start": 458,
            "end": 467,
            "source": {
              "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
              "name": "GraphQL request"
            }
          }
        },
        "name": {
          "kind": "Name",
          "value": "user",
          "loc": {
            "start": 469,
            "end": 473,
            "source": {
              "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
              "name": "GraphQL request"
            }
          }
        },
        "arguments": [

        ],
        "directives": [

        ],
        "selectionSet": {
          "kind": "SelectionSet",
          "selections": [
            {
              "kind": "Field",
              "alias": null,
              "name": {
                "kind": "Name",
                "value": "name",
                "loc": {
                  "start": 496,
                  "end": 500,
                  "source": {
                    "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                    "name": "GraphQL request"
                  }
                }
              },
              "arguments": [

              ],
              "directives": [

              ],
              "selectionSet": null,
              "loc": {
                "start": 496,
                "end": 500,
                "source": {
                  "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                  "name": "GraphQL request"
                }
              }
            },
            {
              "kind": "Field",
              "alias": null,
              "name": {
                "kind": "Name",
                "value": "avatarUrl",
                "loc": {
                  "start": 521,
                  "end": 530,
                  "source": {
                    "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                    "name": "GraphQL request"
                  }
                }
              },
              "arguments": [

              ],
              "directives": [

              ],
              "selectionSet": null,
              "loc": {
                "start": 521,
                "end": 530,
                "source": {
                  "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
                  "name": "GraphQL request"
                }
              }
            }
          ],
          "loc": {
            "start": 474,
            "end": 550,
            "source": {
              "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
              "name": "GraphQL request"
            }
          }
        },
        "loc": {
          "start": 458,
          "end": 550,
          "source": {
            "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
            "name": "GraphQL request"
          }
        }
      }
    ],
    "loc": {
      "start": 401,
      "end": 568,
      "source": {
        "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
        "name": "GraphQL request"
      }
    }
  },
  "loc": {
    "start": 373,
    "end": 568,
    "source": {
      "body": "{\n  checkin(id: 9) {\n    entityId\n    createdAt\n    relationship {\n      manager {\n        ...UserDetails\n      }\n      report {\n        ...UserDetails\n        email\n        company {\n          slug\n        }\n      }\n    }\n    answers {\n      edges {\n        node {\n          body(format: HTML)\n          comments {\n            edges {\n              node {\n                ... on CheckinAnswerComment {\n                  body(format: HTML)\n                  commenter: user {\n                    name\n                    avatarUrl\n                  }\n                }\n              }\n            }\n          }\n          question {\n            body\n          }\n        }\n      }\n    }\n  }\n}\n\nfragment UserDetails on User {\n  name\n  avatarUrl\n}\n",
      "name": "GraphQL request"
    }
  }
}
```

Happy to try and create a more robust solution with a little bit of guidance. But this small change fixed this issue for me.

### CODE REVIEW

**Comment by ekosz** [comment]
@mickhansen Sorry to bother you, but what do you think of this change?

**Comment by mickhansen** [comment]
@ekosz Sorry for slow response! Mind adding a test that showcases this issue?

File: src/simplifyAST.js [hunk 1]
```diff
@@ -23,7 +23,10 @@ function hasFragments(info) {
 }
 
 function isFragment(info, ast) {
-  return hasFragments(info) && info.fragments[ast.name.value] && ast.kind !== 'FragmentDefinition';
+  return hasFragments(info) &&
+    !!ast.name &&
```
Comment (mickhansen) [inline src/simplifyAST.js]: `ast.name` should be sufficient

**Comment by ekosz** [comment]
@mickhansen I really wanted to try out v3/v4 but I still need this fix. I've gone ahead and responded to your comment. Would love to see this in master so I stop running off of my fork.

**Comment by mickhansen** [comment]
@ekosz If you could add a test that would be great. Although v4 no longer relies on AST so unless you use it yourself it shouldn't matter.

**Comment by ekosz** [comment]
Hey @mickhansen I know it's been a while for this PR, but I tried using `4.0.5` of `graphql-sequalize` and I'm still getting this error.

```
GraphQLError: Cannot read property 'value' of undefined
  File "/usr/src/app/node_modules/graphql-sequelize/lib/simplifyAST.js", line 28, in isFragment
    return hasFragments(info) && info.fragments[ast.name.value] && ast.kind !== 'FragmentDefinition';
  File "/usr/src/app/node_modules/graphql-sequelize/lib/simplifyAST.js", line 68, in simplifyAST
    if (isFragment(info, ast)) {
  File "/usr/src/app/node_modules/graphql-sequelize/lib/simplifyAST.js", line 79, in selections.reduce.fields
    simpleAST = deepMerge(simpleAST, simplifyAST(selection, info));
  ?, in Array.reduce
```

I would love the ability to stop having to fork the library. This is a really connor edge case and other than the data I posted above I wouldn't know how to replicate in a test. Would we be able to merge this PR and release as `4.0.6`?

**Comment by mickhansen** [comment]
@ekosz As soon as we get a test tos how the issue and guard against regressions.
